File: Source/communication/src/main/java/protocol/NamedWrapperTranslator.java
package protocol;

import root.communication.ProtocolTranslator;
import root.communication.messages.Message;
import root.communication.messages.MessageType;
import root.communication.parser.DataParser;

public class NamedWrapperTranslator implements ProtocolTranslator {

	private NameTypeResolver typeResolver;
	private DataParser parser;

	public NamedWrapperTranslator(DataParser parser, NameTypeResolver messageTypeResolver) {

		this.parser = parser;
		this.typeResolver = messageTypeResolver;
	}

	@Override
	public String toStrData(Message message) {
		var wrapper = new NamedWrapper(
				message.type.toString(),
				parser.ToString(message));

		return parser.ToString(wrapper);
	}

	@Override
	public byte[] toByteData(Message message) {
		return this.toStrData(message).getBytes();
	}

	@Override
	public Message toMessage(String strData) {

		Message message = null;

		try {
			var wrappedData = parser.FromString(strData, NamedWrapper.class);
			var messageType = typeResolver.resolve(MessageType.valueOf(wrappedData.name));

			if (messageType != null) {
				System.out.println("Resolved message class: " + messageType.toString());
				message = (Message) parser.FromString(wrappedData.payload, messageType);
			} else {
				System.out.println("Unknown message type received: " + wrappedData.name);
			}

		} catch (Exception e) {
			System.out.println("exception while unwrapping/translating/casting received message ... ");
			System.out.println(e.getMessage());
		}

		return message;
	}

	@Override
	public Message toMessage(byte[] byteData) {
		return this.toMessage(new String(byteData));
	}

}

File: Source/server/rabbit-game/src/SharedModel/NamedWrapperTranslator.cs
using RabbitGameServer.SharedModel.Messages;
using RabbitGameServer.Util;

namespace RabbitGameServer.SharedModel
{
	public class NamedWrapperTranslator : IProtocolTranslator
	{

		private ISerializer serializer;

		public NamedWrapperTranslator(ISerializer serializer)
		{
			this.serializer = serializer;
		}

		public byte[] ToByteData(Message message)
		{
			var wrappedMsg = new NamedWrapper(message.type.ToString(),
				serializer.ToString(message));

			return serializer.ToBytes(wrappedMsg);
		}

		public string ToStrData(Message message)
		{
			var wrappedMsg = new NamedWrapper(
				message.type.ToString(),
				serializer.ToString(message));

			return serializer.ToString(wrappedMsg);
		}

		public Message ToMessage(string payload)
		{
			var wrappedMsg = serializer.ToObj<NamedWrapper>(payload);
			Console.WriteLine($"Unwrapping: {wrappedMsg.name}");
			Console.WriteLine("Name: " + wrappedMsg.name);
			Console.WriteLine("Payload: " + wrappedMsg.payload);

			try
			{

				if (wrappedMsg.name == MessageType.ReadyForInitMsg.ToString())
				{
					return serializer.ToObj<ReadyForInitMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.InitializeMessage.ToString())
				{
					return serializer.ToObj<InitializeMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.MoveMessage.ToString())
				{
					return serializer.ToObj<MoveMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.CreateRoomRequest.ToString())
				{
					return serializer.ToObj<CreateRoomMsg>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.JoinRoomRequest.ToString())
				{
					return serializer.ToObj<JoinRoomMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.RoomResponse.ToString())
				{
					return serializer.ToObj<RoomResponseMsg>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.LeaveRoomRequest.ToString())
				{
					return serializer.ToObj<LeaveRoomMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.StartGameRequest.ToString())
				{
					return serializer.ToObj<StartGameMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.AttackMessage.ToString())
				{
					return serializer.ToObj<AttackMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.DefendMessage.ToString())
				{
					return serializer.ToObj<DefendMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.AbortAttackMessage.ToString())
				{
					return serializer.ToObj<AbortAttackMessage>(wrappedMsg.payload);
				}
				else if (wrappedMsg.name == MessageType.AbortDefenseMessage.ToString())
				{
					return serializer.ToObj<AbortDefenseMessage>(wrappedMsg.payload);
				}
				else
				{
					Console.WriteLine("Unwrapping unknown message type ... ");
					Console.WriteLine("Please check your WrapperTranslator ... :)");
				}
			}
			catch (Exception e)
			{
				Console.WriteLine("Exc while casting unwrapped message to concrete type ... ");
				Console.WriteLine(e.Message);

				return null;
			}

			return null;
		}

		public Message ToMessage(byte[] payload)
		{
			return ToMessage(bytesToString(payload));
		}

		private string bytesToString(byte[] someBytes)
		{
			return System.Text.Encoding.Default.GetString(someBytes);
		}

	}
}